onc-healthit/lantern-back-end
# Cache endpoint availability

Added fhir_endpoints_availability table that takes number of http response 200 returned and number of total responses for each endpoint
Add availability column to fhir_endpoints_info
Add migrate up scripts that will add column and table and populate from existing db
Add migrate down script to remove changes

To test (I recommend pgAdmin for this) - 
1. Run and populate the database
2. Check that fhir_endpoints_availability exist and is correctly populated
3. Check that fhir_endpoints_info has availability column. Availability should be 1.0 for endpoints that have 200 status
4. Check that fhir_endpoints_info_history also has this info
5. Change the http_response of endpoints and verify that availability is updated accordingly 

To test migrate scripts: 

1. run and populate the database on master
2. checkout this branch
3. run `make migrate_database`   - to migrate up
4. Use pgadmin to verify that availability table and column were created and populated
5. Change line 42 in `db/migration/cmd/main.go` to be `m.Steps(-2)`
6. run `make migrate_database`  - to migrate down
7. Verify that tables are gone 




Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: https://oncprojectracking.healthit.gov/support/browse/LANTERN-255
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
  - Code accomplishes the tasks purpose.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.
